pkg/speccheck/builtin/process_deployment.go

```diff
@@ -1,27 +1,51 @@
 package builtin
 
 import (
+	"crypto/sha256"
+	"encoding/hex"
+	"hash"
 	"strconv"
 	"strings"
 
 	"github.com/atlassian/smith"
 	"github.com/atlassian/smith/pkg/speccheck"
 	"github.com/atlassian/smith/pkg/util"
+	"github.com/pkg/errors"
 	"go.uber.org/zap"
 	apps_v1 "k8s.io/api/apps/v1"
+	core_v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
 const (
 	// LastAppliedReplicasAnnotation is the name of annotation which stores last applied replicas for deployment
 	LastAppliedReplicasAnnotation = smith.Domain + "/LastAppliedReplicas"
+	EnvRefHashAnnotation          = smith.Domain + "/envRefHash"
 )
 
 type deployment struct {
 }
 
-func (deployment) ApplySpec(ctx *speccheck.Context, spec, actual *unstructured.Unstructured) (runtime.Object, error) {
+func (d deployment) BeforeCreate(ctx *speccheck.Context, spec *unstructured.Unstructured) (runtime.Object /*updatedSpec*/, error) {
+	var deploymentSpec apps_v1.Deployment
+	if err := util.ConvertType(appsV1Scheme, spec, &deploymentSpec); err != nil {
+		return nil, err
+	}
+
+	if deploymentSpec.Annotations == nil {
+		deploymentSpec.Annotations = make(map[string]string)
+	}
+	d.setLastAppliedReplicasAnnotation(ctx, &deploymentSpec, nil)
+	err := d.setConfigurationHashAnnotation(ctx, &deploymentSpec)
```

Comment:
Note how both methods are invoking these functions (see same file below). This achieves same behaviour for both objects that need to be updated and new objects that are created.